storage/raftentry: TestConcurrentUpdates failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/25dd36f0139bf65b80758deeeccf35ee17ebd622

Parameters:
```
TAGS=
GOFLAGS=
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestConcurrentUpdates PKG=github.com/cockroachdb/cockroach/pkg/storage/raftentry TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1260521&tab=buildLog

```
=== RUN   TestConcurrentUpdates
--- FAIL: TestConcurrentUpdates (0.08s)
=== RUN   TestConcurrentUpdates/drop
    --- FAIL: TestConcurrentUpdates/drop (0.03s)
        cache_test.go:484: expected cache to have 0 entries, got 2
        cache_test.go:484: expected cache to have 0 bytes, got 182

```

SHA: https://github.com/cockroachdb/cockroach/commits/a53852a8f6c02ca5573a22abc03c790326ef69ba

Parameters:
```
TAGS=
GOFLAGS=
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestConcurrentUpdates PKG=github.com/cockroachdb/cockroach/pkg/storage/raftentry TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1262650&tab=buildLog

```
=== RUN   TestConcurrentUpdates
--- FAIL: TestConcurrentUpdates (0.05s)
=== RUN   TestConcurrentUpdates/drop
    --- FAIL: TestConcurrentUpdates/drop (0.03s)
        cache_test.go:484: expected cache to have 0 bytes, got 182

```

--- a/pkg/storage/raftentry/cache_test.go
+++ b/pkg/storage/raftentry/cache_test.go
@@ -149,12 +149,20 @@ func TestEntryCache(t *testing.T) {
 
 func verifyMetrics(t *testing.T, c *Cache, expectedEntries, expectedBytes int64) {
 	t.Helper()
+	// NB: Cache gauges are updated asynchronously. In the face of concurrent
+	// updates gauges may not reflect the current cache state. Force
+	// synchrononization of the metrics before using them to validate cache state.
+	c.syncGauges()
 	if got := c.Metrics().Size.Value(); got != expectedEntries {
 		t.Errorf("expected cache to have %d entries, got %d", expectedEntries, got)
 	}
 	if got := c.Metrics().Bytes.Value(); got != expectedBytes {
 		t.Errorf("expected cache to have %d bytes, got %d", expectedBytes, got)
 	}
+}
+
+func (c *Cache) syncGauges() {
+	c.updateGauges(c.addBytes(0), c.addEntries(0))
 }
 
 func TestIgnoredAdd(t *testing.T) {

--- a/pkg/storage/raftentry/metrics.go
+++ b/pkg/storage/raftentry/metrics.go
@@ -45,6 +45,8 @@ var (
 
 // Metrics is the set of metrics for the raft entry cache.
 type Metrics struct {
+	// NB: the values in the gauges are updated asynchronously and may hold stale
+	// values in the face of concurrent updates.
 	Size     *metric.Gauge
 	Bytes    *metric.Gauge
 	Accesses *metric.Counter

